[WIP] Version 0.1

valendesigns opened the pull request:
- [x] REST API Support
- [x] Add test coverage
- [x] Setup Travis CI
- [x] Setup Coveralls
- [x] Create API user flows & validate
- [x] Add README.md
- [ ] Add contributors list

valendesigns committed:
Development version 0.1

valendesigns committed:
Additional configs

valendesigns committed:
Update build matrix

valendesigns committed:
Remove whitespace

valendesigns committed:
Add php-coveralls

valendesigns committed:
Clean up

valendesigns committed:
Fix build

valendesigns committed:
Clean up tools & add tests

valendesigns committed:
Ensure email|login changes invalidate token

valendesigns committed:
Ensure password change revokes all tokens

valendesigns committed:
Only support REST for now

kadamwhite commented on .eslintrc:
```diff
@@ -0,0 +1,156 @@
+{
+	"env": {
+		"browser": true
+	},
```
We'll probably want to clean this up in a similar way to wordpress/grunt-patch-wordpress#72 -- consume the core eslint plugin, etc -- but that's a follow-up, not necessary to resolve for this PR.

kadamwhite replied:
Pulled this out into #3 so we can table it for now and revisit once the important stuff is done :ok_hand:

kadamwhite commented on .travis.yml:
```diff
@@ -0,0 +1,40 @@
+
+# Newer versions like trusty don't have PHP 5.2 or 5.3
+# https://blog.travis-ci.com/2017-07-11-trusty-as-default-linux-is-coming
+dist: precise
```
With the work we're doing to bump minimum version, let's code towards a `7.x` future and only test down to 5.6 at minimum.

valendesigns replied:
We can test down to 5.6, which is fine, but we still would support 5.3 through static analysis.

kadamwhite replied:
I slightly disagree, 5.3 is long past EOL and the timeline we have for this would put us into a zone where we shouldn't have to worry about it with regard to core development. I don't think it should hold this up but I don't think we need to go all the way back to precise.

kadamwhite commented on jwt-auth.php:
```diff
@@ -0,0 +1,72 @@
+
+	if ( version_compare( PHP_VERSION, '5.3', '<' ) ) {
+		deactivate_plugins( plugin_basename( __FILE__ ) );
+		wp_die( esc_html__( 'The JWT Auth plugin requires PHP Version 5.3 or above.', 'jwt-auth' ) );
```
See above note about 5.6

valendesigns replied:
@kadamwhite Changing this to 5.6 is not needed. The minimum PHP version required for the code to work is 5.3 and Core is adopting 5.3, so what would the benefit of making this 5.6 be?

kadamwhite replied:
Per this post: https://wordpress.org/news/2019/04/minimum-php-version-update/ the PHP minimum version for core is, "5.6 now and soon 7+"

kadamwhite commented on jwt-auth.php:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	/** WP_REST_Token Class */
+	require_once JWT_AUTH_PLUGIN_DIR . '/wp-includes/rest-api/auth/class-wp-rest-token.php';
```
PHPDoc question: is `/** */` the appropriate syntax here? I'd expect that `/**` would be used within that file, and this would be `//`, but i'm less familiar with PHPDoc than JSDoc and others.

kadamwhite commented on tests/autoload.php:
```diff
@@ -0,0 +1,28 @@
+$config = getenv( 'WP_TESTS_CONFIG' );
+
+/**
+ * Supports loading the wp-tests-config.php from a non VVV custom directory.
```
Nitpick: let's remove the VVV-specific language.

valendesigns replied:
At this point the only environment the unit tests work in is VVV, or have been tested in would be more accurate.

renatonascalves replied:
@valendesigns Do you mean it can only run inside VVV? Because I was able to run it via Laravel Valet.

renatonascalves replied:
![Screen Shot 2019-03-19 at 09 17 13](https://user-images.githubusercontent.com/19148962/54605347-d341de00-4a27-11e9-9f3d-0d0ffbb0115a.png)

kadamwhite commented on wp-admin/includes/class-wp-key-pair-list-table.php:
```diff
@@ -0,0 +1,139 @@
+				<div class="alignright">
+					<?php submit_button( esc_html__( 'Revoke All', 'jwt-auth' ), 'delete', 'revoke-all-key-pairs', false ); ?>
+				</div>
+			<?php endif; ?>
```
Missing alternative condition for `'top' === $which`?

kadamwhite commented on wp-admin/includes/class-wp-key-pair-list-table.php:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	/**
+	 * Generates content for a single row of the table.
```
This is the same description as `column_default()`: file needs docblock scrub.

kadamwhite commented on wp-admin/js/key-pair.js:
```diff
@@ -0,0 +1,186 @@
+/* global keyPair, confirm, wp */
+( function( $, keyPair ) {
+	'use strict';
```
I'm deferring review on this file because we'll likely want to refactor it to account for modern core JS standards later on.

kadamwhite commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+ *
+ * This class only allows REST authentication to the `{api_prefix}/wp/v2/token`
+ * endpoint with an API key:secret. This allows a user to be identified via the
+ * REST API without using their login credentials to generate a JSON Web Token.
```
I see what's intended here, but as discussed in Slack I think this flow is confusing and that we'll need a request flow example (Either in js, php, whichever, saying, POST to /wp-json/tokens/generate, then take X from that response, etc…)  before we make any progress on that.

valendesigns committed:
Fix spacing

kadamwhite commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+	 */
+	public function register_routes() {
+		$args = array(
+			array(
```
I'm not clear why this is an array of arrays. Shouldn't these all just be one level? None of the core endpoints use nested arrays in this way, they pass `$args = array( 'args' => array(), 'methods' => ... )` etc.

renatonascalves replied:
I second that!

kadamwhite commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+				'callback' => array(
+					$this,
+					'delete_key_pair',
+				),
```
Absolute style nit but we try to keep these `callback` arrays to one line in the core controller code, to make it more obvious it's a class method lookup.

renatonascalves replied:
Second that!

kadamwhite commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+		if ( isset( $_POST['pass1'] ) && ! empty( $_POST['pass1'] ) ) { // phpcs:ignore
+			$keypairs = $this->get_user_key_pairs( $user_id );
+			if ( ! empty( $keypairs ) ) {
+				$this->set_user_key_pairs( $user_id, array() );
```
Could use a clarifying comment, here or in the function docbloc, clearly explaining what's being cleared.

kadamwhite commented on wp-includes/rest-api/auth/jwt/JWT.php:
```diff
@@ -0,0 +1,379 @@
+<?php
+
+namespace Firebase\JWT;
```
Anybody know the policy on external dependencies like this, in terms of where they should go in the codebase? Most of them go into `/wp-includes/[ project ]`, looking at SimplePie, Requests, etc. I'd expect this should be followed here

valendesigns replied:
I've moved it 👍

kadamwhite reviewed (changes requested):
This plugin's a WIP prototype, so I have no formal objection to merging as-is if it'll keep us moving; there's lots to be done and this does provide a decent foundation.

However, there's a couple wishlist things I'd like to see here before pushing the green button:

- A little more rigor around documentation, commenting, and style; on the theory that if we start strong, it'll be easier to end strong
- A comment thoroughly walking through the flow of requests necessary to generate a token
- Some skeleton documentation around the overall structure; as evidenced by the conversation in Slack today several of us are not confidant we can suss out the flows from the code alone, so a talk-through would be very valuable.

jasonbahl commented:
I see in the checklist: `Add support for WPGraphQL`

I think right now, as long as there are appropriate filters/hooks/public methods for _any_ API to hook into, then at the moment it should be WPGraphQL's responsibility to provide some "adapter" to work with this. 

There's no official plan (that I know of) to move WPGraphQL into core, so a plugin like this that _is_ targeting core should probably not be concerned with explicit WPGraphQL support, but rather general access for _any_ API layer to make use of. 

REST and GraphQL aren't the only API's that will exist for WordPress. Whatever cool new API tech 5 years from now should _also_ be able to make use of whatever foundation is there. 

----

I'll try and review the code shortly and provide thoughts on how I would want to hook in and if any changes are necessary to be able to accomplish what I would need to do

valendesigns commented:
@kadamwhite I'll work on updates from the review on Monday. Thanks for reviewing!

@jasonbahl Yes the intent is to ensure the current hooks/filters and workflow support WPGraphQL not to write specific code only used for it.

renatonascalves commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+		$args = array(
+			array(
+				'methods'  => WP_REST_Server::CREATABLE,
+				'callback' => array(
```
Suggestion: leave in one line: `'callback' => array( $this, 'generate_key_pair' )` This and others below.

renatonascalves commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+	public function get_item_schema() {
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => esc_html__( 'Key-pair', 'jwt-auth' ),
```
Shouldn't this be lower case and with the underscore? `key_pair`?

renatonascalves commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+	 * @param mixed           $user    The user that's being authenticated.
+	 * @param WP_REST_Request $request The REST request object.
+	 *
+	 * @return bool|object|mixed
```
Suggestion: Add the objects returned: `WP_User` and `WP_Error` so that I don't have to read the code/method to know which objects are returned.

valendesigns replied:
Technically if someone has filtered `rest_authentication_user` it could be any kind of object. If `$user` is not equal to `false` the method returns the value so yes the intention is that it will return a `WP_User` but that's not a guarantee and will work perfectly with any type of object with the correct properties. If someone wanted to build their own decoupled user system that's possible by filtering `rest_authentication_user`. I'll update the docblock to call this out.

renatonascalves replied:
K! Thank you! Makes sense!

renatonascalves commented on wp-includes/rest-api/auth/class-wp-rest-key-pair.php:
```diff
@@ -0,0 +1,798 @@
+	 *
+	 * @param object $jwt The JSON Web Token.
+	 *
+	 * @return object|WP_Error
```
Same here! Which `object` is that?

renatonascalves reviewed (changes requested):
Thoughts on my initial review!

jkmassel committed:
Allow GET requests to use tokens, if provided

jkmassel committed:
Fix manual token creation with plain permalinks

Uses the built-in `get_rest_url` API to handle all situations.

Return only the path from `get_rest_uri`

jkmassel committed:
Enforce the default permalink style in testing

jkmassel committed:
Test both permalink types

ocean90 committed:
Add missing vendor name in package name

kadamwhite commented:
The key thing we're waiting on here, if you have bandwidth @valendesigns , is this:

> A comment thoroughly walking through the flow of requests necessary to generate a token

If we can get a human-language walkthrough of the full auth process, that will help move everything else along.

ocean90 commented on composer.json:
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "jwt-auth",
```
Created https://github.com/WP-API/jwt-auth/pull/7 to fix the missing namespace.

ocean90 commented:
I'm wondering what the reason was for invalidating a token when an email changes in https://github.com/WP-API/jwt-auth/pull/1/commits/7ccfc84b4cb44c5c2b67f84c4c309a00d946ae61?
This now means when a user changes their email via the API you'd have to request them to re-login after the change to get a valid token again. This sounds like a bad UX and would be different from the existing cookie validation since only a `user_login` change invalidates the cookie.

koke commented:
I’d extend the same question to user_login. I don’t see why a username change would invalidate a token.

I even have doubts about whether a password change should, not sure if there has been a discussion about that, or just going with what cookie authentication does.

ocean90 commented:
A password change doesn't invalidate the token currently, which I think is fine.

Since one of the principles of JWT is stateless I'm suggesting to just drop `validate_user()` altogether. A valid signed and not expired token needs to be trusted. If it's valid let `wp_set_current_user( $user_id );` do the rest.

If we follow this we could remove all the user data/type handling to keep it simple. Maybe even use the [`sub` claim](https://tools.ietf.org/html/rfc7519#section-4.1.2) for the user ID?

jasonbahl commented:
I need to get caught up on the codebase and conversations, but I would submit that Username and password changes _should_ for sure invalidate tokens. 

If we go the "stateless" approach, we need to issue very short lived tokens (few minute expiration max) and an additional revokeable longer-lived refresh token. 

A token that's valid when it's issued should only be trusted for a very short time. 

If a users account is compromised - let's say a malicious co-worker or relative used their username and password - the compromised user should be able to take action, like resetting their password, and the compromised tokens should no longer be valid. 

Even think of Netflix, etc which give you a "Sign out of all accounts" option. That's invalidating all previously issued tokens. 

The way we've approached this on WPGraphQL JWT Authentication is issuing 2 tokens. A short-lived access/auth token (5 min expiration) and a long-lived (1-year) refresh token. 

The authToken can be used to do things like read private data, execute mutations (writes), etc. 

The refreshToken is _only_ good for getting a new `authToken`. If an `authToken` is compromised, it's limited to 5 minutes, which may even be considered long, especially if you're the one who's token was compromised. 

Anyway, our flow looks like this:

- User logs in and asks for a `jwtAuthToken` and `jwtRefreshToken` in repsonse. 
- for future requests, the `jwtAuthToken` is checked to see if it's expired or not. 
   - If _not_ expired, it's added to the headers of the request and used. 
   - If it _is_ expired, a pre-flight request is sent using the `jwtRefreshToken` to ask for a new `jwtAuthToken`
      - If the request succeeds, a new `jwtAuthToken` is provided
      - if the request fails (the `jwtRefreshToken` is invalid. . .manually revoked by an admin, pw reset, etc) then a new `jwtAuthToken` is _not_ issued and the user must login again to get new tokens. 


If we're using a single token (haven't really dug into the code yet) then it absolutely needs to be stateful.

ocean90 commented:
@jasonbahl The standard specification doesn't know about auth/refresh tokens or token invalidation. What you have implemented are additional layers on top of the JWT standard.

jasonbahl commented:
@ocean90 I mean, JWTs are a method of transport, not a comprehensive auth flow. . .what auth flow standard are you referring to so I can get on the same page?

jasonbahl commented:
Regardless, in WordPress, no matter how we do this auth is going to be stateful, because `wp_set_current_user( $id )` already adds the state. It makes a DB request (state) and sets the globals for the user. 

Truly stateless would not even need to make a DB call, and would instead trust the claims in the tokens. . .the roles/caps would be part of the token, and if valid those roles/caps would be used instead of asking the database for the roles/caps of the user.

ocean90 commented:
Good points about the separation of transport and authentication. In this case we probably have to discuss about whether the suggested implementation of auth tokes is practical for real-world apps. See also #8.

ocean90 commented on wp-includes/rest-api/auth/class-wp-rest-token.php:
```diff
@@ -0,0 +1,703 @@
+	public function require_token() {
+		$require_token  = true;
+		$request_uri    = isset( $_SERVER['REQUEST_URI'] ) ? $_SERVER['REQUEST_URI'] : false; // phpcs:ignore
+		$request_method = isset( $_SERVER['REQUEST_METHOD'] ) ? $_SERVER['REQUEST_METHOD'] : false; // phpcs:ignore
```
Created https://core.trac.wordpress.org/ticket/46907 so we don't have to rely on `$_SERVER`.

ocean90 commented on wp-includes/rest-api/auth/class-wp-rest-token.php:
```diff
@@ -0,0 +1,703 @@
+			}
+
+			// Determine if the token has expired.
+			$expiration_valid = $this->validate_expiration( $jwt );
```
This should never be reached if the token has expired because [`JWT::decode()` throws an exception](https://github.com/firebase/php-jwt/blob/v5.0.0/src/JWT.php#L115-L135) in this case.

We should probably catch the specific errors `\Firebase\JWT\BeforeValidException` and `\Firebase\JWT\ExpiredException` before falling back to the generic `\Exception`.

kadamwhite committed:
Merge pull request #7 from ocean90/patch-1

Add missing vendor name in package name

jkmassel committed:
Always verify the token, if provided

Per https://github.com/WP-API/jwt-auth/pull/5#issuecomment-482076262

jkmassel committed:
Merge branch 'develop' into fix/privileged-get-requests-not-working

jkmassel committed:
Create README.md

valendesigns committed:
Update min PHP version & dev-lib

valendesigns committed:
Remove VVV language

valendesigns committed:
Clean up array args

valendesigns committed:
Clarify the removal of key-pairs on password change

valendesigns committed:
Move JWT into correct dependencies directory

valendesigns committed:
Add more docs & clean array args

valendesigns committed:
Exclude JWT file from WPCS

valendesigns committed:
Fix missing xmllint

valendesigns committed:
Use latest linux container

valendesigns committed:
Exclude JWT from tests

valendesigns committed:
Merge pull request #10 from WP-API/review-fixes

Review fixes

valendesigns committed:
Update plugin header

valendesigns committed:
Add refresh tokens

valendesigns committed:
Merge branch 'develop' into refresh-token

valendesigns committed:
Update UI to handle new refresh token

valendesigns committed:
Revert test name change

valendesigns committed:
Merge pull request #11 from WP-API/refresh-token

Refresh Tokens

jkmassel committed:
Fix test failures

jkmassel committed:
Merge branch 'develop' into fix/privileged-get-requests-not-working

jkmassel committed:
Merge branch 'develop' into add-readme

jkmassel committed:
Fix Multi-site

jkmassel committed:
Merge branch 'develop' into fix/manual-tokens-not-working-for-plain-permalinks

jkmassel committed:
Fix token checking

valendesigns committed:
Merge changes & rework solution

valendesigns committed:
Rework docs & ensure token is generated with key-pair only

valendesigns committed:
Merge pull request #9 from WP-API/add-readme

Create readme.md

valendesigns committed:
Add token validate endpoint

valendesigns committed:
Fix spacing

valendesigns committed:
Merge branch 'fix/privileged-get-requests-not-working' of https://github.com/jkmassel/jwt-auth into fix-get-requests

valendesigns committed:
Fix typo

valendesigns committed:
Update GET request comment

valendesigns committed:
Merge branch 'fix-get-requests' into develop

valendesigns committed:
Merge branch 'develop' into auth-fix

valendesigns committed:
Ensure fields autoselect. closes #2

valendesigns committed:
Merge branch 'develop' into auth-fix

valendesigns committed:
Merge pull request #12 from WP-API/auth-fix

Allow for both plain and pretty permalink structures

kadamwhite commented:
After discussion with @TimothyBJacobs , the size of this PR is prohibitive for a thorough review. We've all had a lot of eyes on it over the past few months; let's land this and iterate rather than letting this linger further. Thank you @valendesigns !

renatonascalves commented:
@kadamwhite agree! Cool! :)
